File: src/main/scala/com/besuikerd/autologistics/common/lib/network/SidedMessageHandler.scala
package com.besuikerd.autologistics.common.lib.network

import com.besuikerd.autologistics.AutoLogistics
import cpw.mods.fml.common.network.simpleimpl.{MessageContext, IMessageHandler, IMessage}
import cpw.mods.fml.relauncher.Side
import net.minecraft.entity.player.EntityPlayer
import scala.annotation.switch

trait SidedMessageHandler[REQ <: IMessage, REPLY <: IMessage] extends IMessageHandler[REQ, REPLY] {
  override def onMessage(message: REQ, ctx: MessageContext): REPLY = {
    (ctx.side : @switch) match{
      case Side.CLIENT => onClientMessage(AutoLogistics.proxy.getPlayer(ctx), message, ctx)
      case Side.SERVER => onServerMessage(AutoLogistics.proxy.getPlayer(ctx), message, ctx)
    }
  }

  def onClientMessage(player:EntityPlayer, message:REQ, ctx:MessageContext): REPLY
  def onServerMessage(player:EntityPlayer, message:REQ, ctx:MessageContext): REPLY
}

File: src/main/java/com/besuikerd/autologistics/common/lib/network/SidedMessageHandler.java
package com.besuikerd.autologistics.common.lib.network;

import com.besuikerd.autologistics.AutoLogistics;
import cpw.mods.fml.common.network.simpleimpl.IMessage;
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
import cpw.mods.fml.common.network.simpleimpl.MessageContext;
import net.minecraft.entity.player.EntityPlayer;

public abstract class SidedMessageHandler<REQ extends IMessage, REPLY extends IMessage> implements IMessageHandler<REQ, REPLY>{
    @Override
    public REPLY onMessage(REQ message, MessageContext ctx) {
        switch(ctx.side){
            case CLIENT:
                if(handlesClientSide()){
                    return onClientMessage(AutoLogistics.proxy().getPlayer(ctx), message, ctx);
                }
            case SERVER:
                if(handlesServerSide()){
                    return onServerMessage(AutoLogistics.proxy().getPlayer(ctx), message, ctx);
                }
            default:
                return null;
        }
    }

    public boolean handlesClientSide(){
        return true;
    }

    public boolean handlesServerSide(){
        return true;
    }

    public REPLY onClientMessage(EntityPlayer player, REQ message, MessageContext ctx){ return null; }
    public REPLY onServerMessage(EntityPlayer player, REQ message, MessageContext ctx){ return null; }
}
